src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -1552,18 +1715,181 @@ CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
 }
 
 //
-// General
+// Cryptography
 //
 
-CHIP_ERROR GroupDataProviderImpl::Decrypt(PacketHeader packetHeader, PayloadHeader & payloadHeader,
-                                          System::PacketBufferHandle & msg)
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::SetKey(const ByteSpan & value)
 {
-    (void) packetHeader;
-    (void) payloadHeader;
-    (void) msg;
+    VerifyOrReturnError(value.size() == Crypto::CHIP_CRYPTO_SYMMETRIC_KEY_LENGTH_BYTES, CHIP_ERROR_BUFFER_TOO_SMALL);
+    memcpy(mKey, value.data(), value.size());
     return CHIP_NO_ERROR;
 }
 
+void GroupDataProviderImpl::GroupKeyContext::Clear()
+{
+    memset(mKey, 0, sizeof(mKey));
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::EncryptMessage(MutableByteSpan & plaintext, const ByteSpan & aad,
+                                                                  const ByteSpan & nonce, MutableByteSpan & out_mic) const
+{
+    uint8_t * output = plaintext.data();
+    return Crypto::AES_CCM_encrypt(plaintext.data(), plaintext.size(), aad.data(), aad.size(), mKey, Crypto::kAES_CCM128_Key_Length,
+                                   nonce.data(), nonce.size(), output, out_mic.data(), out_mic.size());
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptMessage(MutableByteSpan & ciphertext, const ByteSpan & aad,
+                                                                  const ByteSpan & nonce, const ByteSpan & mic) const
+{
+    uint8_t * output = ciphertext.data();
+    return Crypto::AES_CCM_decrypt(ciphertext.data(), ciphertext.size(), aad.data(), aad.size(), mic.data(), mic.size(), mKey,
+                                   Crypto::kAES_CCM128_Key_Length, nonce.data(), nonce.size(), output);
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::EncryptPrivacy(MutableByteSpan & plaintext, const ByteSpan & nonce) const
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+CHIP_ERROR GroupDataProviderImpl::GroupKeyContext::DecryptPrivacy(MutableByteSpan & ciphertext, const ByteSpan & nonce) const
+{
+    return CHIP_ERROR_NOT_IMPLEMENTED;
+}
+
+GroupDataProviderImpl::GroupSessionIterator * GroupDataProviderImpl::IterateGroupSessions(uint16_t session_id)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mGroupSessionsIterator.CreateObject(*this, session_id);
+}
+
+GroupDataProviderImpl::GroupSessionIteratorImpl::GroupSessionIteratorImpl(GroupDataProviderImpl & provider, uint16_t session_id) :
+    mProvider(provider), mSessionId(session_id)
+{
+    FabricList fabric_list;
+    ReturnOnFailure(fabric_list.Load(provider.mStorage));
+    mFirstFabric = fabric_list.first_fabric;
+    mFabric      = fabric_list.first_fabric;
+    mFabricCount = 0;
+    mFabricTotal = fabric_list.fabric_count;
+    mMapCount    = 0;
+    mFirstMap    = true;
+}
+
+size_t GroupDataProviderImpl::GroupSessionIteratorImpl::Count()
```

Comment:
Count could be implemented as a basic loop over Next() with save/restore of iterator context around just going through a counting loop of `while (Next())`

Reply:
I gave this a try, but I didn't like the result. Since there is so much state to save and restore, the new function was only eight lines shorter than the current version. The current version is also more efficient since it incurs in less loads, and not need to set the output.